packages/@aws-cdk/cdk/lib/lazy.ts

```diff
@@ -0,0 +1,166 @@
+import { IResolveContext, Token } from "./token";
+
+/**
+ * Interface for lazy string producers
+ */
+export interface IStringValue {
+  /**
+   * Produce the string value
+   */
+  produce(context: IResolveContext): string | undefined;
+}
+
+/**
+ * Interface for lazy list producers
+ */
+export interface IListValue {
+  /**
+   * Produce the list value
+   */
+  produce(context: IResolveContext): string[] | undefined;
+}
+
+/**
+ * Interface for lazy number producers
+ */
+export interface INumberValue {
+  /**
+   * Produce the number value
+   */
+  produce(context: IResolveContext): number | undefined;
+}
+
+/**
+ * Interface for lazy untyped value producers
+ */
+export interface IAnyValue {
+  /**
+   * Produce the value
+   */
+  produce(context: IResolveContext): any;
+}
+
+/**
+ * Options for creating a lazy string token
+ */
+export interface LazyStringValueOptions {
+  /**
+   * Use the given name as a display hint
+   *
+   * @default - No hint
+   */
+  readonly displayHint?: string;
+}
+
+/**
+ * Options for creating a lazy list token
+ */
+export interface LazyListValueOptions {
+  /**
+   * Use the given name as a display hint
+   *
+   * @default - No hint
+   */
+  readonly displayHint?: string;
+
+  /**
+   * If the produced list is empty, return 'undefined' instead
+   *
+   * @default false
+   */
+  readonly omitEmpty?: boolean;
+}
+
+/**
+ * Options for creating lazy untyped tokens
+ */
+export interface LazyAnyValueOptions {
+  /**
+   * Use the given name as a display hint
+   *
+   * @default - No hint
+   */
+  readonly displayHint?: string;
+
+  /**
+   * If the produced value is an array and it is empty, return 'undefined' instead
+   *
+   * @default false
+   */
+  readonly omitEmptyArray?: boolean;
+}
+
+/**
+ * Lazily produce a value
```

Comment:
To me, `Lazy` means that we'll spare a bunch of expensive compute if some value is never used. That's not what we do here, more like deferring the compute to a later stage (the `prepare` or `synth` phase, I guess?).

I don't think `Lazy` is so fundamentally wrong that this is a deal breaker to me, however I kinda feel like something like `Deferred` or `SynthesisTime` would be a more accurate description of what happens.